### PR TITLE
timewarrior 1.0.0 (new formula)

### PR DESCRIPTION
**Pull request by pirj**
- [X] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/homebrew-core/blob/master/.github/CONTRIBUTING.md) document?
- [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [X] Have you built your formula locally prior to submission with `brew install <formula>` (where `<formula>` is the name of the formula you're submitting)?
- [X] Does your submission pass `brew audit --new-formula <formula>` (after doing `brew install <formula>`)?

---

### CODE REVIEW

File: Formula/timew.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
+class Timew < Formula
```
Comment (zmwangx) [inline Formula/timew.rb]: Formula name should be `timewarrior` (i.e., source file should be `timewarrior.rb`), and class name should be `Timewarrior`.

File: Formula/timew.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  desc "Timewarrior is a command-line time tracking application"
```
Comment (zmwangx) [inline Formula/timew.rb]: Description should be "Command-line time tracking application".

File: Formula/timew.rb (hunk 1 above)
Comment (zmwangx) [inline Formula/timew.rb]: By the way, `timew` could be an alias if it's also commonly known as such.

File: Formula/timew.rb [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  homepage "https://taskwarrior.org/docs/timewarrior/"
+  url "http://taskwarrior.org/download/timew-1.0.0.tar.gz"
+  sha256 "ac027910e1e8365bdd218a8b42389b26d017d38d3c96516c408db6d5a44e0bb5"
```
Comment (zmwangx) [inline Formula/timew.rb]: Maybe also `head "https://git.tasktools.org/scm/tm/timew.git"` after this line.

File: Formula/timew.rb [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+  depends_on "cmake" => :build
+
+  def install
+    system "cmake", ".", *std_cmake_args
+    system "make"
+    system "make", "install"
+  end
+
+  test do
+    system "timew"
```
Comment (zmwangx) [inline Formula/timew.rb]: Test is failing, because it shows an interactive Y/N prompt.

Comment (zmwangx) [inline Formula/timew.rb]: Also, just calling the executable isn't a good test. It's best to ask it do something, then somehow validate the output.

**Comment by zmwangx** [comment]
Thanks for submitting a new formula. A couple inline comments. By the way, the commit subject should be `timew 1.0.0 (new formula)` (formulae is the plural form of formula).

**Comment by pirj** [comment]
Thanks for lightning fast reply and remarks. All seems to be fixed, however please take another look, @zmwangx.

File: Formula/timewarrior.rb [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+  depends_on "cmake" => :build
+
+  def install
+    system "cmake", ".", *std_cmake_args
+    system "make"
+    system "make", "install"
+  end
+
+  test do
+    system "timew"
```
Comment (zmwangx) [inline Formula/timewarrior.rb]: This is not addressed yet.

**Comment by zmwangx** [comment]
Thanks for addressing the various problems, but please take a look at the test block.

**Comment by zmwangx** [comment]
By the way, it is not in the checklist, but you should really run `brew test timewarrior` to see it break for yourself.

Re update: `which timew` is not really a good test. Please refer to https://github.com/Homebrew/brew/blob/master/share/doc/homebrew/Formula-Cookbook.md#add-a-test-to-the-formula for guidelines.

**Comment by pirj** [comment]
@zmwangx I did, but that file was already in place, and either the tool itself doesn't detect the shell as non-interactive, either the shell is interactive during tests.
I've addressed this checking against the version number.
Do you want me to add this as a preferable method to [Formula doc](http://www.rubydoc.info/github/Homebrew/brew/master/Formula#test-class_method) instead of current `We are fine if the executable does not error out`?

**Comment by pirj** [comment]
@BrewStatusBot build and test

File: Formula/timewarrior.rb [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+  depends_on "cmake" => :build
+
+  def install
+    system "cmake", ".", *std_cmake_args
+    system "make"
+    system "make", "install"
+  end
+
+  test do
+    assert_equal version, shell_output("timew --version").strip
```
Comment (zmwangx) [inline Formula/timewarrior.rb]: `version.to_s` actually.

**Comment by zmwangx** [comment]
> I've addressed this checking against the version number.

Sorry, I saw the `system "which timew"` commit but not the latest.

> Do you want me to add this as a preferable method to Formula doc instead of current We are fine if the executable does not error out?

Is it really not possible to have a more substantial test? Testing `--version`, `--help` etc. is a fallback when it's hard to test anything else, which is usually reserved for daemon, stuff that require authentication, etc.

**Comment by zmwangx** [comment]
> @BrewStatusBot build and test

The bot will automatically requeue your job when you push to this branch.

**Comment by pirj** [comment]
> The bot will automatically requeue your job when you push to this branch.

@zmwangx I've seen other people [asking BrewTestBot to rebuild](https://github.com/Homebrew/homebrew-core/pull/4598#issuecomment-245463473)

File: Formula/timewarrior.rb (hunk 6 above)
Comment (pirj) [inline Formula/timewarrior.rb]: Redunant according to [docs ("The version string")](http://www.rubydoc.info/github/Homebrew/brew/master/Formula#version-class_method)

**Comment by zmwangx** [comment]
> I've seen other people asking BrewTestBot to rebuild.

You might have noticed the "Homebrew member" badge there 😆 It only works for maintainers. And that's not necessary in most cases, because as I said the bot automatically requeues when you push. When occasionally manually trigger a rebuild when the earlier build failed for whatever reason that doesn't require any code change.

File: Formula/timewarrior.rb (hunk 6 above)
Comment (zmwangx) [inline Formula/timewarrior.rb]: It works, but it's the convention anyway so might as well respect it.

```
$ grep -rI 'assert_match version.to_s' /usr/local/Library/Taps/homebrew/homebrew-core | wc -l
150
```

**Comment by pirj** [comment]
@zmwangx `version` usage changed according to conventions.
Thanks for ramp-up! Do you feel there's anything that needs to be done before merging?

1. Do you want me to address the example section for `test do` ?
2. I feel that interactive/non-interactive might be a common issue, do you want me to take a look if it's a Jenkins issue by some reason?
3. I see several build scheduled after mine, 7395-7401, unfortunately with no details. It seems some of them might be duplicates for the same PR, and make no sense in case of rebase and push --force (-with-lease).

Do you think those three issues deserve their own issues, or are well-known caveats already?

**Comment by ilovezfs** [comment]
```
timewarrior 1.0.0 (new formula)
```
not
```
timewarrior 1.0.0 (new formulae)
```

File: Formula/timewarrior.rb [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+  homepage "https://taskwarrior.org/docs/timewarrior/"
+  url "http://taskwarrior.org/download/timew-1.0.0.tar.gz"
+  sha256 "ac027910e1e8365bdd218a8b42389b26d017d38d3c96516c408db6d5a44e0bb5"
+  head "https://git.tasktools.org/scm/tm/timew.git"
+
+  depends_on "cmake" => :build
+
+  def install
+    system "cmake", ".", *std_cmake_args
+    system "make"
```
Comment (ilovezfs) [inline Formula/timewarrior.rb]: don't need this. just "make", "install" works in this case

**Comment by ilovezfs** [comment]
For the test, two things
1. it needs to use the full path
2. I'd suggest the following to get around the interactivity issues, though you could do something more meaningful if you'd like
```
test do
    (testpath/".timewarrior/data").mkpath
    (testpath/".timewarrior/extensions").mkpath
    touch ".timewarrior/timewarrior.cfg"
    assert_match "Tracking foo", shell_output("#{bin}/timew start foo")
end
```

**Comment by zmwangx** [comment]
> Do you want me to address the example section for test do ?

What do you mean by "the example section"? If you mean a better test I'm all for it.

> I feel that interactive/non-interactive might be a common issue, do you want me to take a look if it's a Jenkins issue by some reason?

It's not a Jenkins issue, happened for me locally too. I didn't have time to investigate, but you might be able to just pipe a yes to it, or better yet, figure out some actual commands that work non-interactively.

> I see several build scheduled after mine.

Each push resulted in a build, no worries. Should be green now.

Heads up, I may not be available over the next few hours.

**Comment by zmwangx** [comment]
Looks like ILZ already has a better test for you.

**Comment by pirj** [comment]
@ilovezfs Thanks a lot for sane example (read: writing test for me).
Make+make install combined, commit message orpho error fixed. Looking forward for the build to pass.

**Comment by zmwangx** [comment]
I took the liberty to edit the PR title so that it matches the final product better. There are lots of queued jobs right now so please be patient.

**Comment by ilovezfs** [comment]
@pirj you're welcome.

**Comment by zmwangx** [comment]
@ilovezfs Happy to :shipit:?

**Comment by DomT4** [comment]
The URL should be `https://`. Feel free to do in post, though.

**Comment by zmwangx** [comment]
@DomT4 Of course, thanks for noticing that. Easy to fix that when actually shipping.

**Comment by zmwangx** [comment]
@DomT4 By the way, IIRC you were planning to audit HTTPS URLs, right?

**Comment by DomT4** [comment]
No worries. It's something that I'm planning to audit sooner rather than later; I'm working on replacing myself robotically.

**Comment by ilovezfs** [comment]
@DomT4 not sure auditing that is going to really be possible because you'll have to be careful to make sure the content actually resolves to the homepage not just that the server claims to accept https and then returns something which may or may not be the homepage. That's easier said than done.

**Comment by zmwangx** [comment]
@ilovezfs Shouldn't be hard to let a bot crawl and update everything and have a human review all the changes?

**Comment by ilovezfs** [comment]
@zmwangx yeah that's more likely to work unless we want to be dealing with bogus brew audit failures frequently.

**Comment by zmwangx** [comment]
Oops I take back my word.

```
$ grep -rI http: | wc -l
2741
```

WTF.

**Comment by ilovezfs** [comment]
I think the following have to be true to insist on https

1. https is not simply redirecting to http.

2. https and http return the same content modulo http/https string differences in the content. If they do not return the same content then https must resolve and http must not resolve.

3. The certificate chain is valid. If the certificate is not valid, then it is not an audit failure. (Lots of servers are configured in such a way that they appear to be claiming to accept https but don't actually accept it.)

**Comment by zmwangx** [comment]
bot.brew.sh :trollface:

**Comment by ilovezfs** [comment]
@zmwangx https://github.com/Homebrew/homebrew.github.io/issues/98

**Comment by ilovezfs** [comment]
>Oops I take back my word.

What's the problem? We'll just have every maintainer do 1 per day for a year

**Comment by zmwangx** [comment]
> We'll just have every maintainer do 1 per day for a year

I think vszakats will beat the said maintainer to it.

**Comment by ilovezfs** [comment]
LOL

**Comment by ilovezfs** [comment]
>@ilovezfs Happy to :shipit:?

@zmwangx :boat:

**Comment by zmwangx** [comment]
Merged as 0fa7093. Thank you for your contribution to Homebrew!

**Comment by pirj** [comment]
Thanks guys! That was an amazing experience ❤️

**Comment by ilovezfs** [comment]
@pirj thanks for the cool new formula!

**Pull request closed without merging**